### Data Cleaning

Load the datasets using pandas

In [1]:
import pandas as pd
import numpy as np

weather = pd.read_csv('../data/raw/weather_daily.csv', 
                            na_values=['NA', ''])
soil = pd.read_csv('../data/raw/soil_sensor_data.csv', 
                              na_values=['NA', ''])
params = pd.read_csv('../data/raw/crop_zone_parameters.csv', 
                                  na_values=['NA', ''])


print("Crop Zone Parameters:")
print(params.head())
print(params.info())

print("\nSoil Sensor Data:")
print(soil.head())
print(soil.info())

print("\nWeather Daily:")
print(weather.head())
print(weather.info())

# Identify specific missing values
print("Missing values in soil sensor:")
print(soil[soil.isna().any(axis=1)])

print("\nMissing values in weather:")
print(weather[weather.isna().any(axis=1)])

# Describe statistics to look for outliers
print("\nSoil sensor summary statistics:")
print(soil.describe())

print("\nWeather daily summary statistics:")
print(weather.describe())

# Check for sensor status issues
print("\nSensor status unique values and counts:")
print(soil['sensor_status'].value_counts())
print("\nRows with sensor status not OK:")
print(soil[soil['sensor_status'] != 'OK'])

# Let's inspect the temperature outlier
print("Temperature outlier row:")
print(weather[weather['temperature_c'] > 35])

# Let's inspect the tank level outlier
print("\nTank level outlier row:")
print(soil[soil['tank_level_liters'] > 5000])

# Let's check for any other zero-flow or weird combinations
print("\nPump drawing power but zero flow:")
print(soil[(soil['pump_power_watts'] > 100) & (soil['pump_flow_lpm'] == 0)])

--- Crop Zone Parameters ---
  zone_id crop_type  area_m2  min_moisture_pct  target_moisture_pct  \
0  Zone_A    tomato      120                22                   33   
1  Zone_B      kale       90                24                   35   
2  Zone_C     maize      180                20                   31   

   field_capacity_pct  drainage_coefficient  
0                  41                  0.18  
1                  43                  0.15  
2                  40                  0.22  
<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   zone_id               3 non-null      str    
 1   crop_type             3 non-null      str    
 2   area_m2               3 non-null      int64  
 3   min_moisture_pct      3 non-null      int64  
 4   target_moisture_pct   3 non-null      int64  
 5   field_capacity_pct    3 non-null      int64  
 6   

Data cleaning and anomaly detection are critical initial steps in data analysis. 
Using **pandas**, we can easily profile our data to find missing values, 
outliers, unit inconsistencies, and sensor anomalies.

### 1. Identifying Missing Values

Missing values occur when no data is recorded for a specific observation. Pandas 
represents missing numerical or categorical data as `NaN` (Not a Number) or 
`None`.

* **`df.isna()` or `df.isnull()**`: Returns a boolean DataFrame where `True` 
indicates a missing value.
* **`df.isna().sum()`**: Counts the number of missing values in each column.
* **Boolean Indexing**: Used to filter and display the exact rows containing 
missing values.

**Examples in the Data**

* In `soil_sensor_data.csv`, row 16 contains a missing value (`NaN`) for 
`soil_moisture_pct` on `2026-03-06`.
* In `weather_daily.csv`, `rainfall_mm` is missing on `2026-03-08`, and 
`humidity_pct` is missing on `2026-03-21`.


### 2. Identifying Outliers

Outliers are data points that deviate significantly from the rest of the 
dataset. They can represent true extreme events or data entry errors.

* **`df.describe()`**: Generates summary statistics (mean, standard deviation, 
min, max, and quartiles). If the `max` or `min` value is vastly different from 
the $75\text{th}$ or $25\text{th}$ percentiles, an outlier is likely present.
* **Interquartile Range (IQR) Method**: Outliers are identified as values 
falling outside the range:

$$[Q_1 - 1.5 \times IQR, \quad Q_3 + 1.5 \times IQR]$$

where $Q_1$ is the $25\text{th}$ percentile, $Q_3$ is the $75\text{th}$ 
percentile, and $IQR = Q_3 - Q_1$. You can compute these using 
`df['column'].quantile(0.25)` and `df['column'].quantile(0.75)`.
* **Z-Score Method**: Measures how many standard deviations ($\sigma$) a data 
point is from the mean ($\mu$):

$$Z = \frac{x - \mu}{\sigma}$$

Typically, any point with $|Z| > 3$ is flagged as an outlier.



**Examples in the Data**

* In `weather_daily.csv`, running `.describe()` shows a maximum `temperature_c` 
of $45.8^\circ\text{C}$ on `2026-03-14`, whereas the $75\text{th}$ percentile 
is only $26.5^\circ\text{C}$. This indicates an extreme outlier.
* In `soil_sensor_data.csv`, the `tank_level_liters` column hits a maximum value 
of $9900\text{ L}$ on the same day (`2026-03-14`), while the normal operating 
capacity stays between $3300\text{ L}$ and $4800\text{ L}$.


### 3. Identifying Inconsistent Units

Inconsistent units occur when measurements are recorded using different systems 
(e.g., mixing Liters and Gallons, or Celsius and Fahrenheit) or when raw data 
scales change unexpectedly.

* **`df.dtypes`**: Ensures all columns possess the expected numerical or 
categorical data type. Text mixed into a numerical column will force the column 
dtype to be `object`.
* **Domain Validation/Value Ranges**: Checking if values fall outside of 
physically or logically possible ranges for the given unit.
* **Cross-Dataset Ratios**: Analyzing whether metrics scale correctly across 
zones or boundaries.


**Examples in the Data**

* The $45.8^\circ\text{C}$ temperature observation could be a unit mix-up. If 
it were accidentally recorded in Fahrenheit, 
$45.8^\circ\text{F} \approx 7.7^\circ\text{C}$ (which would represent a sudden 
cold snap). If it was meant to be $25.8^\circ\text{C}$, it represents a typo.
* The $9900\text{ L}$ tank level reading might also be an uncalibrated raw 
value or an unscaled metric that missed a decimal point (e.g., 
$4900\text{ L}$ or $3900\text{ L}$).

### 4. Identifying Sensor Anomalies

Sensor anomalies are caused by hardware malfunctions, transmission errors, or 
conflicting operational telemetry (e.g., a device reporting that it is drawing 
power but producing zero output).

* **Categorical Filtering**: Filtering explicit error flags or maintenance 
statuses built into the data schema.
* **Logical Contradiction Rules**: Using conditional boolean indexing (with 
logical operators `&` and `|`) to catch physically impossible combinations 
between multiple sensors.

**Examples in the Data**
* In `soil_sensor_data.csv` at row 61 (`2026-03-21`), the `sensor_status` is 
explicitly flagged as `'CHECK'`.
* By applying cross-sensor logic, pandas reveals that on this row, 
`pump_power_watts` is $468\text{ W}$ (the pump is fully active) but 
`pump_flow_lpm` is $0.0\text{ lpm}$ (no water is moving). This contradiction 
efficiently isolates a hardware anomaly (e.g., a blocked pipe, dry run, or 
broken flow meter).

### Descriptive Statistics of Key Variables

In [4]:
weather_clean = pd.read_csv('../data/processed/weather_daily_cleaned.csv')
soil_clean = pd.read_csv('../data/processed/soil_sensor_data_cleaned.csv')
crop_zone = pd.read_csv('../data/processed/crop_zone_parameters_cleaned.csv')

print("Weather Cleaned Stats:")
print(weather_clean.describe())

print("\nSoil Sensor Cleaned Stats:")
print(soil_clean.describe())

print("\nCrop Zone Stats:")
print(crop_zone.describe())


Weather Cleaned Stats:
       rainfall_mm  temperature_c  humidity_pct  wind_speed_mps  solar_index
count    30.000000      30.000000     30.000000       30.000000    30.000000
mean      8.268333      24.551667     65.566667        2.089333     0.675333
std      16.154152       1.800310      4.073322        0.510564     0.096552
min       0.000000      21.400000     56.300000        1.140000     0.490000
25%       0.662500      23.175000     63.425000        1.660000     0.582500
50%       2.600000      24.700000     64.850000        2.115000     0.695000
75%       8.000000      26.225000     68.250000        2.397500     0.760000
max      85.000000      27.400000     74.600000        3.100000     0.800000

Soil Sensor Cleaned Stats:
       soil_moisture_pct  tank_level_liters  pump_flow_lpm  pump_power_watts
count          90.000000          90.000000      90.000000         90.000000
mean           27.555000        4123.644444      21.893333        459.044444
std             4.551256 

```
```
 
These metrics are computed from the cleaned version of the data so that the 
previously identified sensor anomalies and outliers do not skew the summary 
results.

**1. Weather Daily Metrics ($N = 30$ days)**

These statistics capture the macro-environmental conditions influencing the crop 
zones over a 30-day period.

| Metric | Rainfall ($mm$) | Temperature ($^\circ\text{C}$) | Humidity ($\%$) | Wind Speed ($m/s$) | Solar Index |
| --- | --- | --- | --- | --- | --- |
| **Mean** | $8.27$ | $24.55$ | $65.57$ | $2.09$ | $0.68$ |
| **Std Dev** | $16.15$ | $1.80$ | $4.07$ | $0.51$ | $0.10$ |
| **Min** | $0.00$ | $21.40$ | $56.30$ | $1.14$ | $0.49$ |
| **25% ($Q_1$)** | $0.66$ | $23.18$ | $63.43$ | $1.66$ | $0.58$ |
| **50% (Median)** | $2.60$ | $24.70$ | $64.85$ | $2.12$ | $0.70$ |
| **75% ($Q_3$)** | $8.00$ | $26.23$ | $68.25$ | $2.40$ | $0.76$ |
| **Max** | $85.00$ | $27.40$ | $74.60$ | $3.10$ | $0.80$ |

The temperature stays within a narrow, stable range ($21.4^\circ\text{C}$ to 
$27.4^\circ\text{C}$) after removing the erroneous $45.8^\circ\text{C}$ reading. 
Rainfall is highly skewed, with a median of only $2.6\text{ mm}$ but a maximum 
downpour day of $85.0\text{ mm}$.

**2. Soil & Irrigation Sensor Metrics ($N = 90$ observations across 3 zones)**

These metrics summarize the soil environment and water delivery operations.

| Metric | Soil Moisture ($\%$) | Tank Level (Liters) | Pump Flow ($lpm$) | Pump Power (Watts) |
| --- | --- | --- | --- | --- |
| **Mean** | $27.56$ | $4123.64$ | $21.89$ | $459.04$ |
| **Std Dev** | $4.55$ | $372.84$ | $3.51$ | $31.51$ |
| **Min** | $8.50$ | $3369.00$ | $15.20$ | $394.00$ |
| **25% ($Q_1$)** | $25.00$ | $3842.75$ | $18.83$ | $434.25$ |
| **50% (Median)** | $27.60$ | $4136.50$ | $21.85$ | $461.00$ |
| **75% ($Q_3$)** | $30.68$ | $4422.00$ | $25.08$ | $488.75$ |
| **Max** | $36.10$ | $4836.00$ | $28.80$ | $523.00$ |

Correcting the $9900\text{ L}$ tank level anomaly brings the maximum true tank 
capacity reading to $4836\text{ L}$ with a standard deviation of 
$372.8\text{ L}$, showing normal cyclic depletion and refilling. Pump flows 
consistently hover around an active rate of $\approx 22\text{ lpm}$ when drawing 
$\approx 459\text{ W}$ of power.

**3. Crop Zone Management Baselines ($N = 3$ zones)**

This metadata highlights the static target limits assigned to the fields 
(`Zone_A`: Tomato, `Zone_B`: Kale, `Zone_C`: Maize).

| Variable | Mean | Standard Deviation | Min | Max |
| --- | --- | --- | --- | --- |
| **Area ($m^2$)** | $130.00$ | $45.83$ | $90.00$ | $180.00$ |
| **Minimum Moisture Threshold ($\%$)** | $22.00$ | $2.00$ | $20.00$ | $24.00$ |
| **Target Moisture Threshold ($\%$)** | $33.00$ | $2.00$ | $31.00$ | $35.00$ |
| **Field Capacity ($\%$)** | $41.33$ | $1.53$ | $40.00$ | $43.00$ |
| **Drainage Coefficient** | $0.183$ | $0.035$ | $0.150$ | $0.220$ |

Across fields, the average target moisture content is $33\%$. Comparing this to 
the **Soil Sensor Mean Moisture** ($27.56\%$), the crops are generally kept 
slightly below their optimal targets on average, but safely above the average 
minimum critical threshold ($22\%$). However, the absolute minimum soil moisture 
recorded was $8.5\%$, implying that certain zones have experienced severe 
underwatering periods that warrant closer monitoring.

### Scientific Visualizations and Interpretations

The five scientific visualizations below analyze the interaction between the 
soil, crop physiology, pump hydromechanical infrastructure, and atmospheric 
microclimate.

### 1. Soil Moisture Dynamics vs. Crop Management Thresholds

#### **Scientific Interpretation:**

This multi-panel visualization tracks the temporal changes in soil moisture 
against three critical agronomic thresholds: **Field Capacity** (the maximum 
water a soil profile can retain against gravity), **Target Moisture** (the 
optimal level for crop metabolic function), and **Minimum Critical Moisture** 
(the threshold below which plants cannot easily extract water, initiating 
stomatal closure).

Together, the zone between the Minimum Critical line and Field Capacity defines 
the **Readily Available Water (RAW)** buffer.

* **Zone A (Tomato):** Displays highly effective micro-irrigation management. 
The observed moisture curve is tightly regulated near the $33\%$ target line, 
avoiding both anaerobic root stress (near Field Capacity) and matric potential 
stress (near the minimum line). This steady state prevents conditions like 
blossom-end rot or fruit splitting, which are common in tomatoes subjected to 
moisture oscillations.
* **Zone B (Kale):** Reveals a critical agronomic failure between March 10th 
and March 20th. The soil moisture collapsed far below the minimum critical 
threshold ($24\%$), plummeting to a hyper-arid $\approx 85\%$. This indicates a 
prolonged breach of the **Management Allowed Depletion ( MAD )** limit. During 
this period, the soil matric suction exceeded the root pulling force, inducing 
severe turgor loss, stomatal cavity collapse, and restricted vegetative biomass 
accumulation.
* **Zone C (Maize):** Shows a progressive, linear drying trend during the first 
half of the month, crossing briefly beneath the critical minimum line ($20\%$) 
before recovering. This indicates that as the Maize crop developed its leaf area 
index (LAI), its daily crop water requirement ($CWR$) outpaced the irrigation 
delivery rate, causing a temporary soil water deficit.


### 2. Pump Hydromechanical Performance Curves

#### **Scientific Interpretation:**

This plot evaluates the thermodynamic and fluid mechanics performance of the 
irrigation infrastructure by mapping Volumetric Flow Rate ($Q$, in Liters per 
Minute) against Electrical Power Consumption ($P$, in Watts). According to 
centrifugal pump affinity laws and system head curves, power draw is directly 
proportional to the product of volumetric flow and the 
**Total Dynamic Head (TDH)**:


$$P \propto Q \times \text{TDH}$$

* **System Segmentation:** The pumps are highly segregated by zone. Zone C 
operates at the highest volumetric discharge ($24 \text{ to } 28 \text{ lpm}$) 
and highest electrical load ($460 \text{ to } 520 \text{ W}$). This is a direct 
engineering response to field sizing; Zone C has the largest surface area (
    $180 \text{ m}^2$) and requires a higher volumetric dose per irrigation 
    cycle.
* **Negative Operational Slopes:** Within each individual zone, the regression 
trendlines exhibit a distinct *negative slope*—as power consumption increases, 
the flow rate decreases. This indicates a variable hydraulic head system. When 
the source water supply tank empties, the elevation head increases, forcing the 
pump to work harder against gravity (drawing higher wattage/torque) while 
xperiencing a corresponding drop in volumetric output.


### 3. Microclimatic Evaporative Demand Drivers (VPD Proxies)

#### **Scientific Interpretation:**

This multivariate plot models the environmental factors that dictate the 
**Vapor Pressure Deficit (VPD)**—the driving force behind crop transpiration 
and soil evaporation. VPD represents the difference between the saturation 
vapor pressure ($e_s$, an exponential function of temperature) and the actual 
atmospheric vapor pressure ($e_a$, driven by relative humidity).

* **Thermodynamic Couplings:** The visualization highlights a clear inverse 
relationship between temperature and humidity. The critical region lies in the 
lower-right quadrant (ambient temperatures $>26^\circ\text{C}$ and relative 
humidity $<62\%$).
* **Aerodynamic Control:** The data points in this high-demand quadrant are 
dominated by large markers (high wind speeds) and deep orange/red hues (high 
solar index). Mechanically, intense solar radiation drives the stomatal 
openings of the plants and warms the canopy, while high wind speeds reduce the 
boundary-layer aerodynamic resistance surrounding the leaves by sweeping away 
saturated air. This combination creates a microclimatic spike in reference 
evapotranspiration ($ET_0$). If irrigation scheduling is purely time-based 
rather than weather-adaptive, crops exposed to this cluster of conditions will 
experience immediate atmospheric drought stress.


### 4. Soil Hydrodynamics: Ecosystem-Scale Response to Precipitation

#### **Scientific Interpretation:**

This visualization captures the macro-level hydrologic mass balance of the 
agricultural ecosystem, governed by the continuity equation:


$$\frac{dS}{dt} = P + I - ET - D$$


where $\frac{dS}{dt}$ is the change in soil water storage, $P$ is meteoric 
precipitation, $I$ is artificial irrigation, $ET$ is evapotranspiration, and 
$D$ is deep percolation/drainage.

* **Hydraulic Recharge Phase:** On March 5th, a major precipitation event 
($\approx 24\text{ mm}$) drives an immediate, vertical spike in ecosystem-scale 
mean soil moisture. This represents a rapid wetting front moving through the 
unsaturated soil zone, filling macro- and micro-pores. A secondary cluster of 
rainfall events around March 10th sustains this high water storage state.
* **The Master Depletion Curve:** Following March 11th, a 12-day meteorological 
dry spell occurs ($P = 0$). During this window, the mean soil moisture exhibits 
a classic exponential decay curve. The steep initial descent (March 12–15) 
represents the gravity-dominated drainage phase ($D$), where water exceeding 
field capacity rapidly percolates below the root zone. The flattening of the 
curve (March 16–23) represents the transition to an evapotranspiration-dominated 
phase ($ET$), where water is drawn out more slowly through root extraction and 
capillary action.


### 5. Crop-Specific Moisture Distribution & Physiological Compliance

#### **Scientific Interpretation:**

This structural box plot evaluates the operational risk and stability of the 
agricultural management system by mapping the full monthly distribution of soil 
moisture for each crop type relative to its specific biological boundaries.

* **Zone A (Tomato - High Compliance):** Represents a highly controlled system. 
The entire interquartile range (IQR) and whiskers are completely contained 
within the safe operational envelope between the minimum critical threshold and 
field capacity. The median sits almost perfectly on the target line ($33\%$). 
This tight variance ensures that the tomato plants never cross the threshold 
into vegetative or reproductive stress.
* **Zone B (Kale - Structural Failure):** Displays extreme operational variance. 
The lower quartile and extended bottom whisker show that for over $25\%$ of the 
month, the crop was subjected to severe under-watering. The median itself sits 
below the target line, indicating that the irrigation system for Zone B 
consistently failed to deliver the required water volume to match the crop's 
transpiration demand.
* **Zone C (Maize - Chronic Deficit):** Shows a highly compressed, low-variance 
distribution, but one that is poorly positioned. The entire box is squeezed into 
the narrow margin directly above the critical minimum line ($20\%$), well below 
the optimal target line ($31\%$). While the system successfully prevented the 
maize from entering acute wilting stress (no long lower tail), keeping the soil 
chronically near the lower boundary restricts nutrient diffusion in the root 
zone, limiting the crop's biomass accumulation and yield potential.